File: Day3/day3.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	inputLines := readLineByLine("input.txt")

	sumErrors := 0
	for _, line := range inputLines {
		sumErrors += int(getPriority(findError(line)))
	}
	fmt.Println("[Part1] Sum of priorities:", sumErrors)

	sumRecurring := 0

	for i := 0; i < len(inputLines); i += 3 {
		recurring := findRecurringInGroup([]string{
			inputLines[i],
			inputLines[i+1],
			inputLines[i+2],
		})

		sumRecurring += int(getPriority(recurring))
	}

	fmt.Println("[Part2] Sum of all groups:", sumRecurring)

}

func getPriority(letter byte) byte {
	if letter > 96 {
		return letter - 96
	}
	return letter - 38
}

func getUniqueAscii(text string) []byte {
	uniqueLetters := map[byte]struct{}{}
	for _, letter := range []byte(text) {
		uniqueLetters[letter] = struct{}{}
	}
	bytes := []byte{}
	for key := range uniqueLetters {
		bytes = append(bytes, key)
	}
	return bytes
}

func findRecurringInGroup(group []string) byte {
	occurences := map[byte]int{}
	for _, compartment := range group {
		compartmentUnique := getUniqueAscii(compartment)
		for _, letter := range compartmentUnique {
			if occurence, ok := occurences[letter]; ok {
				occurences[letter] = occurence + 1
			} else {
				occurences[letter] = 1
			}
		}
	}

	return getThreeOccurenceLetter(occurences)
}

func getThreeOccurenceLetter(occurences map[byte]int) byte {
	for key, value := range occurences {
		if value == 3 {
			return key
		}
	}
	panic("could not find 3 times recurring letter in a group")
}

func findError(compartment string) byte {
	occurence := map[byte]struct{}{}
	asciiList := []byte(compartment)
	half := len(asciiList) / 2

	for i := 0; i < len(asciiList); i++ {
		codeUnit := asciiList[i]
		if i < half {
			occurence[codeUnit] = struct{}{}
		} else if _, ok := occurence[codeUnit]; ok {
			return codeUnit
		}
	}

	panic("could not find any recurring letter")
}

func readLineByLine(filePath string) []string {
	readFile, err := os.Open(filePath)

	if err != nil {
		fmt.Println(err)
	}
	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	fileLines := []string{}
	for fileScanner.Scan() {
		fileLines = append(fileLines, fileScanner.Text())
	}

	readFile.Close()

	return fileLines
}

File: Day3/day3.py
file = open('input.txt', 'r')
input_lines = file.readlines()

def get_priority(letter):
    return letter - 96 if letter > 96 else letter - 38

def find_error(compartment):
    occurence = set()
    half = len(compartment) / 2

    for i in range(len(compartment)):
        code_unit = ord(compartment[i])
        if i < half:
            occurence.add(code_unit)
        elif code_unit in occurence:
            return code_unit

    raise Exception("could not find any recurring letter");

def get_unique_ascii(text):
    unique_letters = set()
    for letter in text:
        unique_letters.add(letter)

    return unique_letters

def find_recurring_in_group(group):
    occurences = dict()

    for compartment in group:
        compartment_unqiue = get_unique_ascii(compartment.strip())
        for letter in compartment_unqiue:
            occurence = occurences.get(letter, 0)
            occurences[letter] = occurence + 1

    return get_three_occurence_letter(occurences)

def get_three_occurence_letter(map):
    for key in map:
        if map[key] == 3:
            return key

    raise Exception("could not find 3 times recurring letter in a group")

sum_errors = 0
for line in input_lines:
    line = line.strip()
    sum_errors += get_priority(find_error(line))

print("[Part1] Sum of priorities:", sum_errors)

sum_recurring = 0
for i in range(0, len(input_lines), 3):
    recurring = find_recurring_in_group([
        input_lines[i],
        input_lines[i + 1],
        input_lines[i + 2],
    ])
    sum_recurring += get_priority(ord(recurring))

print("[Part2] Sum of all groups:", sum_recurring)